src/target/riscv/riscv.c

```diff
@@ -5029,6 +5040,73 @@ int riscv_dmi_write_u64_bits(struct target *target)
 	return r->dmi_write_u64_bits(target);
 }
 
+static int check_if_trigger_is_availible(struct target *target, unsigned int t)
+{
+	/* If we can't write tselect, then this hart does not support triggers. */
+	if (riscv_set_register(target, GDB_REGNO_TSELECT, t) != ERROR_OK)
+		return ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
+	riscv_reg_t tselect_rb;
+	if (riscv_get_register(target, &tselect_rb, GDB_REGNO_TSELECT) != ERROR_OK)
+		return ERROR_FAIL;
+	/* Mask off the top bit, which is used as tdrmode in old
+	 * implementations. */
+	tselect_rb &= ~(1ULL << (riscv_xlen(target) - 1));
+	if (tselect_rb != t)
+		return ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
+	return ERROR_OK;
+}
+
+static int get_trigger_types(struct target *target, unsigned int *trigger_tinfo,
+		riscv_reg_t tdata1)
+{
+	riscv_reg_t tinfo;
+	const bool may_support_multiple_trigger_types =
+		riscv_get_register(target, &tinfo, GDB_REGNO_TINFO) == ERROR_OK;
+	if (may_support_multiple_trigger_types) {
+		/* tinfo == 0: invalid
+		 * tinfo == 1: trigger doesn’t exist */
+		if (tinfo == 0 || tinfo == 1)
+			return ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
```

Comment:
I know this is the original code, just moved. But if you like, you can be more defensive here: you can end the enumeraton if the LSB of tinfo is set, regardless of the other bits. This is to defend against potential buggy HW implementations of tinfo.

```suggestion
		 * tinfo has LSB set: trigger doesn’t exist */
		if (tinfo == 0 || tinfo & 0x1)
			return ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
```

Reply:
Is this really beneficial to implement workarounds for an implementation that does not behave as specified in advance? I'm afraid this adds unnecessary complexity. If there is a HW implementation that needs a workaround, I would like to see it applied as a separate commit with a solid reasoning and a reproduction. IMHO this makes the intention clearer and results in a more supportable codebase.

Reply:
Defensive approach is about pro-actively finding and fixing points which can be made safer - in advance, not waiting until something breaks.

In this case the implementation is trivial (and can be clearly described in a comment), so I don't mind making it as part of the change. However, I don't insist either.

Actually, the `tinfo == 0` part is already a similar situation - it is a case that the debug spec does not address, but the implementation in OpenOCD already checks for it.

Reply:
Addressed.